src/query_plan/mod.rs

```diff
@@ -1 +1,210 @@
+    node: PlanNode,
+}
+
+/// A `DeferNode` corresponds to one or more `@defer` applications at the same level of "nestedness"
+/// in the planned query.
+///
+/// It contains a "primary block" and a vector of "deferred blocks". The "primary block" represents
+/// the part of the query that is _not_ deferred (so the part of the query up until we reach the
+/// @defer(s) this handles), while each "deferred block" correspond to the deferred part of one of
+/// the @defer(s) handled by the node.
+///
+/// Note that `DeferNode`s are only generated if defer support is enabled for the query planner.
+/// Also note that if said support is enabled, then `DeferNode`s are always generated if the query
+/// has a @defer application, even if in some cases generated plan may not "truly" defer the
+/// underlying fetches (i.e. in cases where `deferred[*].node` are all undefined). This currently
+/// happens because some specific cases of defer cannot be handled, but could later also happen if
+/// we implement more advanced server-side heuristics to decide if deferring is judicious or not.
+/// This allows the executor of the plan to consistently send a defer-abiding multipart response to
+/// the client.
+pub struct DeferNode {
+    /// The "primary" part of a defer, that is the non-deferred part (though could be deferred
+    /// itself for a nested defer).
+    primary: PrimaryDeferBlock,
+    /// The "deferred" parts of the defer (note that it's a vector). Each of those deferred elements
+    /// will correspond to a different chunk of the response to the client (after the initial
+    /// on-deferred one that is).
+    deferred: Vec<DeferredDeferBlock>,
+}
+
+/// The primary block of a `DeferNode`.
+pub struct PrimaryDeferBlock {
+    /// The part of the original query that "selects" the data to send in that primary response
+    /// once the plan in `node` completes). Note that if the parent `DeferNode` is nested, then it
+    /// must come inside the `DeferredNode` in which it is nested, and in that case this
+    /// sub-selection will start at that parent `DeferredNode.query_path`. Note that this can be
+    /// `None` in the rare case that everything in the original query is deferred (which is not very
+    /// useful  in practice, but not disallowed by the @defer spec at the moment).
+    sub_selection: Option<SelectionSet>,
+    /// The plan to get all the data for the primary block. Same notes as for subselection: usually
+    /// defined, but can be undefined in some corner cases where nothing is to be done in the
+    /// primary block.
+    node: Option<PlanNode>,
+}
+
+/// A deferred block of a `DeferNode`.
+pub struct DeferredDeferBlock {
+    /// References one or more fetch node(s) (by `id`) within `DeferNode.primary.node`. The plan of
+    /// this deferred part should not be started until all such fetches return.
+    depends: Vec<DeferredDependency>,
+    /// The optional defer label.
+    label: Option<NodeStr>,
```

Comment:
Is this really optional? Isn’t there code to add labels to `@defer` directive applications that don’t already have one?

Reply:
The defer normalization within query planning makes sure that every `@defer` application without a label has a unique label. However, we explicitly do not include such assigned labels in the query plan ([relevant code here](https://github.com/apollographql/federation/blob/f69a0694b95e99c22f0d5f7bc6df1b46568b2a8d/query-planner-js/src/buildPlan.ts#L3607)). FWIW I'm not sure this is bug-free code, but this is currently the code that's there.